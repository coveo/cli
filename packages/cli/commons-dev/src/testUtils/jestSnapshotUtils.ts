// See https://github.com/oclif/core/blob/f996086397158b5beafc4254718c78fd93d23b1b/src/errors/errors/cli.ts#L55-L60
export const formatCliLog = (input: string) =>
  input.replaceAll(/\r\n/gm, '\n').replaceAll(/»/gm, '›');

export const formatAbsolutePath = (input: string) =>
  input.replaceAll(/\w:\\{2}/gm, '/').replaceAll(/\\{2}/gm, '/');

export const removeDotCmd = (input: string) => input.replaceAll(/\.cmd/gm, '');
