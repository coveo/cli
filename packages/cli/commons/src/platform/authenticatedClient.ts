import {setGlobalDispatcher, ProxyAgent, FormData, fetch} from 'undici';
import PlatformClient from '@coveo/platform-client';
import {Config, Configuration} from '../config/config';
import {castEnvironmentToPlatformClient} from './environment';
import globalConfig from '../config/globalConfig';
import {lt} from 'semver';

export class AuthenticatedClient {
  public cfg: Config;
  public constructor() {
    this.cfg = new Config(globalConfig.get().configDir);
  }

  public async isLoggedIn() {
    const {accessToken} = this.cfg.get();
    return accessToken !== undefined;
  }

  public async isExpired() {
    try {
      const c = await this.getClient();
      await c.initialize();
      return false;
    } catch (e) {
      return true;
    }
  }

  public async getClient(overrideConfig?: Partial<Configuration>) {
    const configFromDisk = this.cfg.get();
    const resolvedConfig = {...configFromDisk, ...overrideConfig};
    const globalRequestSettings: Record<string, unknown> = {};
    const proxyServer =
      process.env['https_proxy'] || process.env['HTTPS_PROXY'];
    if (proxyServer) {
      setGlobalDispatcher(new ProxyAgent(proxyServer));
    }
    if (lt(process.version, '18.0.0')) {
      Object.assign(global, {FormData, fetch});
    }
    return new PlatformClient({
      globalRequestSettings,
      environment: castEnvironmentToPlatformClient(resolvedConfig.environment),
      region: resolvedConfig.region,
      organizationId: resolvedConfig.organization,
      accessToken: resolvedConfig.accessToken!,
    });
  }

  public async getAllOrgsUserHasAccessTo() {
    const platformClient = await this.getClient();
    return platformClient.organization.list();
  }

  public async getUserHasAccessToOrg(org: string) {
    const orgs = await this.getAllOrgsUserHasAccessTo();
    return orgs.some((o) => o.id === org);
  }

  public async createImpersonateApiKey(name: string) {
    const platformClient = await this.getClient();
    return await platformClient.apiKey.create({
      displayName: `cli-${name}`,
      description: 'Generated by the Coveo CLI',
      enabled: true,
      privileges: [
        {targetDomain: 'IMPERSONATE', targetId: '*', owner: 'SEARCH_API'},
      ],
    });
  }

  public async getUsername() {
    const {anonymous, organization} = this.cfg.get();
    if (anonymous) {
      return `api-key@${organization}`;
    }
    const authenticatedClient = new AuthenticatedClient();
    const platformClient = await authenticatedClient.getClient();
    await platformClient.initialize();

    return (await platformClient.user.get()).providerUsername;
  }
}

export enum AuthenticationStatus {
  LOGGED_IN,
  EXPIRED,
  LOGGED_OUT,
}

export async function getAuthenticationStatus() {
  const authenticatedClient = new AuthenticatedClient();

  if (!(await authenticatedClient.isLoggedIn())) {
    return AuthenticationStatus.LOGGED_OUT;
  }

  if (await authenticatedClient.isExpired()) {
    return AuthenticationStatus.EXPIRED;
  }

  return AuthenticationStatus.LOGGED_IN;
}
