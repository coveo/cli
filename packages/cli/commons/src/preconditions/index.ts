export * from './authenticated';
export * from './apiKeyPrivilege';
export * from './preconditions';
