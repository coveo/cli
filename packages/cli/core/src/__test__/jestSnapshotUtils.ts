import strip from 'strip-ansi';

// See https://github.com/oclif/core/blob/f996086397158b5beafc4254718c78fd93d23b1b/src/errors/errors/cli.ts#L55-L60
export const formatCliLog = (input: string, stripAinsi = false) => {
  const formatted = input.replaceAll(/\r\n/gm, '\n').replaceAll(/»/g, '›');
  return stripAinsi ? strip(formatted) : formatted;
};
