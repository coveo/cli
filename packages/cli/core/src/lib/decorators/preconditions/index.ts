// TODO: Remove, unusued
export * from './git';
// TODO: Move to `@coveo/plugin-cli-ui
export * from './node';
// TODO: Move to `@coveo/plugin-cli-ui
export * from './npx';
// TODO: Move to `@coveo/plugin-cli-ui
export * from './npm';
// TODO: Move to `@coveo/plugin-cli-ui
export * from './ng';
