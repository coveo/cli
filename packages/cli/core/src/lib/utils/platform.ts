import {Config} from '@coveo/cli-commons/config/config';

export function getTargetOrg(config: Config, target?: string) {
  if (target) {
    return target;
  }
  const cfg = config.get();
  return cfg.organization;
}
