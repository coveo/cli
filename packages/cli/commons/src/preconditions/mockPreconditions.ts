import {CLICommand} from '../command/cliCommand';
import {PreconditionError} from '../errors/preconditionError';
import {DecoratorFunction} from '../decorators/decoratorFunction';

const thrower = (reason: string) => {
  throw new PreconditionError(`${reason} Precondition Error`);
};

export const mockPreconditions = <
  PreconditionStatus extends Record<string, boolean>
>(
  preconditionStatus: PreconditionStatus
) => {
  type PreconditionKeys = keyof PreconditionStatus & string;
  type PreconditionPromises = Record<PreconditionKeys, DecoratorFunction>;

  const keys: PreconditionKeys[] = Object.keys(preconditionStatus);
  const mockedPreconditions: Partial<PreconditionPromises> = {};
  for (const key of keys) {
    mockedPreconditions[key] = (_target?: CLICommand) =>
      new Promise<void>((resolve) =>
        preconditionStatus[key] ? resolve() : thrower(key)
      );
  }

  return mockedPreconditions as PreconditionPromises;
};
