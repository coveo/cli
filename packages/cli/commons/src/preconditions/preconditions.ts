import {Before} from '../decorators/before';
import { DecoratorFunction } from '../decorators/decoratorFunction';

/**
 * @deprecated use @type {DecoratorFunction}
 */
export type PreconditionFunction = DecoratorFunction;

/**
 * @deprecated use @function {Before}
 */
export const Preconditions = Before;
