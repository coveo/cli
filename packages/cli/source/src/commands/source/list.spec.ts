jest.mock('@coveo/cli-commons/analytics/amplitudeClient');
jest.mock('@coveo/cli-commons/config/config');
jest.mock('@coveo/cli-commons/preconditions/authenticated');

jest.mock('@coveo/cli-commons/platform/authenticatedClient');
jest.mock('@coveo/platform-client');

import {test} from '@oclif/test';
import {AuthenticatedClient} from '@coveo/cli-commons/platform/authenticatedClient';
import {
  SourceModel,
  SourceStatusType,
  SourceType,
  SourceVisibility,
} from '@coveo/platform-client';
import {IsAuthenticated} from '@coveo/cli-commons/preconditions';
import {mockPreconditions} from '@coveo/cli-commons/preconditions/mockPreconditions';
import {formatCliLog} from '@coveo/cli-commons-dev/testUtils/jestSnapshotUtils';

describe('source:push:list', () => {
  const mockedAuthenticatedClient = jest.mocked(AuthenticatedClient);
  const mockedIsAuthenticated = jest.mocked(IsAuthenticated);

  const mockListSources = jest.fn();

  const createMockSourceModel = (id: string): SourceModel => ({
    name: `${id}_displayName`,
    id,
    owner: 'bob',
    sourceVisibility: SourceVisibility.SECURED,
    sourceType: SourceType.PUSH,
    information: {
      sourceStatus: {type: SourceStatusType.PUSH_READY},
      numberOfDocuments: 1234,
    },
  });

  const doMockListSource = function () {
    mockListSources.mockReturnValue(
      Promise.resolve({totalEntries: 0, sourceModels: []})
    );
  };

  const doMockPreconditions = function () {
    const preconditionStatus = {
      authentication: true,
    };
    const mockedPreconditions = mockPreconditions(preconditionStatus);
    mockedIsAuthenticated.mockReturnValue(mockedPreconditions.authentication);
  };

  beforeEach(() => {
    mockedAuthenticatedClient.mockImplementation(
      () =>
        ({
          cfg: {
            get: () => ({organization: 'foo'}),
          },
          getClient: () => Promise.resolve({source: {list: mockListSources}}),
        } as unknown as AuthenticatedClient)
    );
    doMockPreconditions();
    doMockListSource();
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  test
    .stdout()
    .stderr()
    .command(['source:list'])
    .it('works when there is no push source configured', (ctx) => {
      expect(formatCliLog(ctx.stdout)).toMatchSnapshot();
    });

  test
    .do(() => {
      mockListSources.mockReturnValueOnce(
        Promise.resolve({
          totalEntries: 1,
          sourceModels: [createMockSourceModel('the_id')],
        })
      );
    })
    .stdout()
    .stderr()
    .command(['source:list'])
    .it('works when the user has access to a list of push source', (ctx) => {
      expect(formatCliLog(ctx.stdout)).toMatchSnapshot();
    });
});
