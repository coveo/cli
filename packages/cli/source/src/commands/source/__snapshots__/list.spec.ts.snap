// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`source:push:list works when the user has access to a list of push source 1`] = `
" Id     Name               Owner Source type Source visibility Status     Number of documents 
 ────── ────────────────── ───── ─────────── ───────────────── ────────── ─────────────────── 
 the_id the_id_displayName bob   PUSH        SECURED           PUSH_READY 1234                
"
`;

exports[`source:push:list works when there is no push source configured 1`] = `
"There is no push nor catalog source in organization foo
You can:
  • create a push source using source:push:new
  • create a catalog source using source:catalog:new
"
`;
